runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java

```diff
@@ -626,52 +628,119 @@ public synchronized long find(long globalAddress, SearchDirection direction) {
+     * that CP2.logicalCheckpointedSpace > CP1.logicalCheckpointedSpace.
+     * Consider the case where, CP2 snapshot was taken before CP1, however, CP1 read/write is faster
+     * and commits its entries to the log first.
+     *
+     * +------------------------------------------------+
+     * | CP1 (snapshot 15) |  |  |  | CP2 (snapshot 10) |
+     * +------------------------------------------------+
+     *
+     * @param context this stream's current context
+     * @param data checkpoint log data entry
+     * @param maxGlobal maximum global address to resolve this stream up to.
+     *
+     * @return true, if the checkpoint was completely resolved (from end to start markers of a checkpoint)
+     *         false, otherwise.
+     */
+    protected boolean scanCheckpointStream(final QueuedStreamContext context, ILogData data,
+                                           long maxGlobal) {
         if (data.hasCheckpointMetadata()) {
-            CheckpointEntry cpEntry = (CheckpointEntry)
-                    data.getPayload(runtime);
+            CheckpointEntry cpEntry = (CheckpointEntry) data.getPayload(runtime);
 
-            // Select the latest cp that has a snapshot address
-            // which is less than maxGlobal
-            if (context.checkpointSuccessId == null &&
+            // Consider only checkpoints that are less than maxGlobal
+            // Because we are traversing in reverse order END marker of a checkpoint should be found first.
+            if (context.checkpoint.checkpointId == null &&
                     cpEntry.getCpType() == CheckpointEntry.CheckpointEntryType.END
                     && Long.decode(cpEntry.getDict()
                     .get(CheckpointEntry.CheckpointDictKey.START_LOG_ADDRESS)) <= maxGlobal) {
                 log.trace("Checkpoint[{}] END found at address {} type {} id {} author {}",
                         this, data.getGlobalAddress(), cpEntry.getCpType(),
-                        Utils.toReadableId(cpEntry.getCheckpointId()),
-                        cpEntry.getCheckpointAuthorId());
-                context.checkpointSuccessId = cpEntry.getCheckpointId();
+                        Utils.toReadableId(cpEntry.getCheckpointId()), cpEntry.getCheckpointAuthorId());
+
+                // Because checkpoint ordering is not guaranteed, i.e., a checkpoint for a lower snapshot
+                // could appear in the log after a checkpoint for a higher snapshot (case of multiple
+                // checkpointers running in parallel). We need to inspect all checkpoints and keep the one
+                // with the highest VLO version.
+                UUID checkpointId = cpEntry.getCheckpointId();
+                long checkpointVLOVersion = Long.decode(cpEntry.getDict()
+                        .get(CheckpointEntry.CheckpointDictKey.START_LOG_ADDRESS));
+                boolean isCheckpointForHighestVLOVersion = latestValidCheckpoint.validateHighest(checkpointId,
```

Comment:
In name `isCheckpointForHighestVLOVersion` and `validateHighest`, `Highest` is confusing here, I think it should be `Higher` right?

Reply:
Sure, if it's more intuitive. Also I added comments to the method.